File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
'''
BFS를 통한 풀이
입구가 출구인 경우 X
입구나 출구가 벽인 경우 X
=>항상 도착위치로 이동 가능
'''
import sys
from collections import deque

# 가능한 범위
def is_valid(x,y):
    return 0<=x<M and 0<=y<N

# 4방 탐색
delta_x = [0,0,-1,1]
delta_y = [1,-1,0,0]

N,M = map(int, input().split())
graph = []
visited = []
for _ in range(N):
    graph.append(list(map(int,list(sys.stdin.readline().rstrip()))))
    visited.append([False]*M)

deq = deque()
deq.append([0,0,1]) #x y time(길이)
visited[0][0] = True
while deq:
    cur = deq.popleft()
    time = cur[2]
    if cur[0] == M-1 and cur[1] == N-1: # 도착
        print(time)
        break
    for k in range(4):
        dx = cur[0] + delta_x[k]
        dy = cur[1] + delta_y[k]
        if is_valid(dx,dy) and graph[dy][dx] == 1 and not visited[dy][dx]: # 가능한 범위, 길, 방문 X
            deq.append([dx,dy,time+1])
            visited[dy][dx] = True

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer token;
    static StringBuilder sb = new StringBuilder();

    static int n,m;
    static boolean[][] graph, visited;
    static int[] deltaX = {0,0,-1,1};
    static int[] deltaY = {1,-1,0,0};

    public static void main(String[] args) throws IOException {
        token = new StringTokenizer(input.readLine());
        n = Integer.parseInt(token.nextToken());
        m = Integer.parseInt(token.nextToken());

        graph = new boolean[n][m];
        visited = new boolean[n][m];

        for (int i = 0; i < n; i++) {
            String s = input.readLine();
            for (int j = 0; j < m; j++) {
                if (s.charAt(j)=='1'){
                    graph[i][j] = true;
                }
            }
        }

        Deque<int[]> deq = new ArrayDeque<>();
        deq.push(new int[]{0,0,1});
        visited[0][0] = true;
        flag:
        while (!deq.isEmpty()) {
            int[] cur = deq.pollFirst();
            int x = cur[0];
            int y = cur[1];
            int cnt = cur[2];

            for (int i=0; i<4; i++){
                if (isValid(x+deltaX[i],y+deltaY[i]) && graph[y+deltaY[i]][x+deltaX[i]] && !visited[y+deltaY[i]][x+deltaX[i]]){
                    if (x+deltaX[i] == m-1 && y+deltaY[i] == n-1){
                        System.out.println(cnt+1);
                        break flag;
                    }
                    deq.addLast(new int[]{x+deltaX[i],y+deltaY[i],cnt+1});
                    visited[y+deltaY[i]][x+deltaX[i]] = true;
                }
            }
        }
    }
    public static boolean isValid(int x, int y) {
        return x>=0 && x<m && y>=0 && y<n;
    }

}